### ISSUE TITLE
Support Iceberg branching for all the snapshot producer operations

### ISSUE THREAD

**Issue by jackye1995** [apache/iceberg#3896]
With the introduction of branching, we can produce snapshots to a specific branch like the following:

```java
table.newAppend()
    .appendFile(someFile)
    .branch("some_branch")
    .commit();
```

**Comment by flyrain** [comment]
I guess people can use branches for testing some things without impact the trunk. The trunk should still be there for any read in that case. Would you mind share any solid use cases? Like in which scenario this would be pretty useful?

**Comment by jackye1995** [comment]
> I guess people can use branches for testing new changes without impact the trunk
yes that is one primary use case. Another one is for a separated branch that maintains a historical snapshot. A compacted version could be updated to the branch for optimized read performance.

**Comment by flyrain** [comment]
Do you mean "a compact version could be updated to the **trunk** for optimized read"? Make sense to me.

**Comment by RussellSpitzer** [comment]
Are we going to allow merging a branch back into the main line? I was thinking about this for transaction support.

**Comment by flyrain** [comment]
IIRC, we got a discussion on this, the verdict is not to support merging a branch into the trunk.

**Comment by amogh-jahagirdar** [comment]
There is also some more clarity with what it could look like based on the discussion in https://github.com/apache/iceberg/pull/3883 (in particular https://github.com/apache/iceberg/pull/3883#discussion_r795274832). @jackye1995 I am happy to take this issue if possible!

**Comment by namratha2403** [comment]
@jackye1995 @amogh-jahagirdar I have been working working on the above issue and have a raised a PR for same. Any comments or suggestions appreciated. PR is:  https://github.com/apache/iceberg/pull/4926

**Comment by syun64** [comment]
Hi folks, I see this PR was merged to enable support for writing to branches with table identifiers in Spark 3.3. https://github.com/apache/iceberg/pull/6965

Do we need to go through a separate effort to backport it to Spark 3.2, and is it possible to backport to 3.2? (Are there any Spark features that are only available in 3.3 that's required for introducing custom table identifiers?)
